cmake/modules/AddSwift.cmake

```diff
@@ -198,38 +201,69 @@ function(_add_variant_c_compile_flags)
     RESULT_VAR_NAME result)
 
   is_build_type_optimized("${CFLAGS_BUILD_TYPE}" optimized)
-  if(optimized)
-    list(APPEND result "-O2")
 
-    # Add -momit-leaf-frame-pointer on x86.
-    if("${CFLAGS_ARCH}" STREQUAL "i386" OR "${CFLAGS_ARCH}" STREQUAL "x86_64")
-      list(APPEND result "-momit-leaf-frame-pointer")
+  # MSVC has different options to most compilers!
+  if("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
+    list(APPEND result "-DLLVM_ON_WIN32")
+    list(APPEND result "-D_CRT_SECURE_NO_WARNINGS")
+    # TODO(compnerd) handle /MT
+    list(APPEND result "-D_DLL")
+
+    if(optimized)
+      list(APPEND result "/O2")
+
+      # Add -momit-leaf-frame-pointer on x86.
+      if("${CFLAGS_ARCH}" STREQUAL "x86" OR "${CFLAGS_ARCH}" STREQUAL "x86_64")
+        list(APPEND result "/Oy")
+      endif()
+    else()
+      list(APPEND result "/Od")
     endif()
   else()
-    list(APPEND result "-O0")
+    if(optimized)
+      list(APPEND result "-O2")
+
+      # Add -momit-leaf-frame-pointer on x86.
+      if("${CFLAGS_ARCH}" STREQUAL "i386" OR "${CFLAGS_ARCH}" STREQUAL "x86_64")
+        list(APPEND result "-momit-leaf-frame-pointer")
+      endif()
+    else()
+      list(APPEND result "-O0")
+    endif()
```

Comment:
I haven't split this out, but I've tried to make this a bit more readable - let me know if you want me to split it out: I found it was too complicated and slightly confusing as you have to pass multiple variables etc. etc. Maybe that's because I'm not a CMake expert tho...